src/Entity/Organization/Package.php

```diff
@@ -246,4 +256,44 @@ private function setName(string $name): void
 
         $this->name = $name;
     }
+
+    /**
+     * @return Collection<int,Version>|Version[]
+     */
+    public function versions(): Collection
+    {
+        return $this->versions;
+    }
+
+    public function addOrUpdateVersion(Version $version): void
+    {
+        if ($this->getVersion($version->version()) !== false) {
+            $this->getVersion($version->version())->setReference($version->reference());
+
+            return;
+        }
+
+        $version->setPackage($this);
+        $this->versions->add($version);
+    }
+
+    /**
+     * @return Version|false
+     */
+    public function getVersion(string $versionString)
+    {
+        return $this->versions->filter(fn (Version $version) => $version->version() === $versionString)->first();
+    }
+
+    /**
+     * @param string[] $encounteredVersions
+     */
+    public function removeUnencounteredVersions(array $encounteredVersions): void
```

Comment:
what about put this logic to `syncSuccess` method? Now this method will be always called before `syncSuccess`, so i think we can add this `$encounteredVersions` param there.

Reply:
Maybe even pass the whole packages to syncsuccess and move the adding/removing logic fully there?

Reply:
make sense

Reply:
So given the below idea about storage: either we need to pass $diskStorage to `syncSuccess` or not move the whole packages there. I think the second option is better? Since otherwise we are passing services.

Reply:
Passing these services is not bad. Generally, we will be doing a slight change of architecture towards a more layered (domain, application, infrastructure) to get away from this symfony style. At the moment it can stay as it is :+1: